halo2/src/prover.rs

```diff
@@ -219,32 +206,19 @@ impl Halo2Prover {
         Ok(proof)
     }
 
-    pub fn assert_field_is_compatible<F: FieldElement>() {
-        if polyexen::expr::get_field_p::<Fr>() != F::modulus().to_arbitrary_integer() {
-            panic!("powdr modulus doesn't match halo2 modulus. Make sure you are using Bn254");
-        }
-    }
-
-    fn verification_key_inner<F: FieldElement>(
-        &self,
-        pil: &Analyzed<F>,
-        fixed: &[(String, Vec<F>)],
-    ) -> Result<VerifyingKey<G1Affine>, String> {
-        let plaf_circuit = analyzed_to_plaf(pil, fixed);
-        let circuit = analyzed_to_circuit_with_zeroed_witness(plaf_circuit, pil);
-        match keygen_vk(&self.params, &circuit) {
-            Ok(vkey) => Ok(vkey),
-            Err(e) => Err(e.to_string()),
-        }
+    pub fn add_verification_key(&mut self, mut vkey: &mut dyn io::Read) {
+        let vkey = VerifyingKey::<G1Affine>::read::<&mut dyn io::Read, PlafH2Circuit>(
+            &mut vkey,
+            SerdeFormat::Processed,
+            self.plaf.clone(),
+        )
+        .unwrap();
+        self.vkey = Some(vkey);
     }
 
-    pub fn verification_key<F: FieldElement>(
-        &self,
-        pil: &Analyzed<F>,
-        fixed: &[(String, Vec<F>)],
-    ) -> Result<Vec<u8>, String> {
-        let vk = self.verification_key_inner(pil, fixed)?;
-        Ok(vk.to_bytes(SerdeFormat::Processed))
+    pub fn verification_key(&self) -> Result<VerifyingKey<G1Affine>, String> {
```

Comment:
Maybe this could already check if we have one and simply clone that?

Reply:
This mimics the behavior of what is in main. `verification_key()` always creates a new key, and if this key is not manually added to the backend, it panics when verifying.

Doesn't look good. Maybe if a key is needed, either by this function or by `verify()`, a new one should be created and set. Or either, create the key if none is supplied in `create()`.

Should I change this in this PR?

Reply:
The key is deterministic so I don't think we need to create a new one. But either is fine (here or another PR)